### ISSUE TITLE
Contribute to Scapy: a master list of good first issues

### ISSUE THREAD

**Issue by gpotter2** [secdev/scapy#399]
Hi & welcome to Scapy's github ! This page lists issues that you can try to fix if you want to start contributing to Scapy.

## Still here?

This list includes wishes and things added by the maintainers based on the issues that we get, but also issues marked with `TODO` or `XXX` that already exist in Scapy's code base (layers). If you want to contribute to the project you might just take care one of the bugs.

~~You can also look for issues and pull requests in the [old Bitbucket repository](https://bitbucket.org/secdev/scapy). If you find something interesting there (there shouldn't be anything left, but sometimes lost people add issues there), please do not update it on Bitbucket, open an issue here and mention (link) the old issue and/or pull request.~~

**To get a bug, create an issue with the description of the bug and refer to this issue (just mention `#399` in the text) so we can update it.**

**Please read [How to contribute](/secdev/scapy/blob/master/CONTRIBUTING.md) before starting to work!**

*Note: all of those issues are updated regularly but might sometimes fall out of date. The links are static so the part you need to change might move on master (look for it). If the issue has been fixed and we missed it, feel free to ping us*

### **Easy - Recommended for new users - Hacktoberfest**
- **Any of the [issues tagged with "good first issue"](https://github.com/secdev/scapy/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)**

- **Participate in project Hinty: https://github.com/secdev/scapy/issues/2158. This is easy but requires a lot of hands to help :smile:**. You'll just have to add static typing to a file or layer you've selected.

- Add documentation for `dispatch_hook` based on [this StackOverflow post](https://stackoverflow.com/a/28025620/5459467). You can add this documentation to `scapy/doc/scapy/advanced_usage.rst` for instance.

- Add `hashret()/answers()` regression tests to `ICMPv6ND_RA` in regression.uts as said
https://github.com/secdev/scapy/blob/45c59302cefa364f6584710f9116f8492e3bdcc6/test/regression.uts#L1692 . This is a good way to discover what our unit test infrastructure looks like (it's a bit outdated though very complete.. we're trying to switch to pytest....)

### **Medium - More advanced features or wishes**
- create a snap package for Scapy (#2153) (apparently that's something Linux users want)

- Implement TCP_server like TCP_client (#2083), except it should handle multiple streams ! (Implement Transmission Control Block? This also leads to understanding how Scapy automatons work)

- Improve `in6_getAddrType`: add support for `Unique local Address` and add unit tests. (this function appears to be unused, try pinging guedou about that)
https://github.com/secdev/scapy/blob/c6e1e6ccd201664db20068c0aaabba83082ea449/scapy/utils6.py#L191

- Add more tests in regression.uts for `IA_PD` Prefix Option in 
https://github.com/secdev/scapy/blob/45c59302cefa364f6584710f9116f8492e3bdcc6/test/regression.uts#L3264

- Make the FlagField a bit cleaner thanks to the new capabilities of `FlagsField` in `isakmp.py` (This should be easy enough, try `help(FlagsField)` to see what this field is about)
https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/layers/isakmp.py#L185

- Don't remove Teredo's route when refreshing in `route6.py`
https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/route6.py#L41 (I'm honnestly not sure what this is about. Ask us :p)

### **Hard - Build entire modules or refactor complex stuff**

- add SNMPv3 support #1132

- add DTLS support #1931 (This is very cool, but not easy. We do have full TLS support though so it should be feasible to re-use almost everything)

- Update the p0f support #1923 (This sounds pretty cumbersome...)

- Remove the "trick" from `carp.py`
https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/contrib/carp.py#L47 (You very likely need to get back to the RFC to understand what the heck is going on here. At least this should be rather simple)

- Try to call `Enum_metaclass` from `ASN1_Class_metaclass` or at least try to reduce the amount of duplicated code:
https://github.com/secdev/scapy/blob/a6e4ab2a45fe5e5bb9b4b15f78ebc4f1778be6b3/scapy/asn1/asn1.py#L113

- Update the OPC UA support #1569 (This sounds extremely hard, considering how big the PR is. Don't dive into this unless you actually need OPC UA)

- Add support to addresses other than MAC in `DUID_LLT` in `dhcp6.py`
https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/layers/dhcp6.py#L204

- Add support for authdata in `OSPF_Hdr` in `ospf.py`
https://github.com/secdev/scapy/blob/f912538b7959a68ad83e7f18eee57cd5a8aef7df/scapy/contrib/ospf.py#L68

- Guess the next hop properly when multiple `IPv6ExtHdrDestOpt` extensions are used on `inet6.py`
https://github.com/secdev/scapy/blob/bdcbba58b87bbd79ac416c9614bcf2f5e9847764/scapy/layers/inet6.py#L379 (You might want to check the RFC: should we use the one from the last option? or the first one? (current))

- Check that Multicast scopes match when performing `answers` in [inet6.py]
https://github.com/secdev/scapy/blob/bdcbba58b87bbd79ac416c9614bcf2f5e9847764/scapy/layers/inet6.py#L407

- Write X.400 format in `X509.py`. (this is ASN.1. Not the easiest thing to begin with)
https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/layers/x509.py#L199

**HELP:** the RFC of the the X.400 format is on [rfc5280](https://tools.ietf.org/html/rfc5280), at the bottom of page 117.

### **Unknown difficulty**
- Fix a bug with ICMPv6's hashret in inet6.py
https://github.com/secdev/scapy/blob/35f97efc53839f6366918c251dba5df2ac553c8d/scapy/layers/inet6.py#L424 (this looks like a bug but it's not crystal clear... 😕)

**Comment by gpotter2** [comment]
@p-l- Should this be in a like CONTRIBUTIONS.md file ? Or do you want to copy paste it and create it again  with your own account so you can update the list ?

**Comment by xee5ch** [comment]
Thank you so much.  This is awesome work?  Perhaps we can build a list of help-wanted issues, and tag them with difficulty levels?  This will be a crazy week, but I am willing to pitch in!

**Comment by gpotter2** [comment]
Yeah good idea: this is the raw list, so maybe we should reorganize it a bit...

**Comment by xee5ch** [comment]
@gpotter2 I am more than willing to start with bug triage while I get acquainted.  @p-l- I kind of stepped on your toes last time inquiring about minor bugfix and cleanup stuff, so I do not want to just jump in unless I have permission: can I move all these to different  issues and try to work with any of you to label them properly?

If not I will begin trying to digest the easier issues and work up to my first constructive patch.

**Comment by p-l-** [comment]
Here is what I propose to add at the end of the list:

If you want to contribute, and have chosen an item in the list on which you want to work: if there is an issue mentioned in the list, use that issue to communicate with other people and see what is going on. If not, open an issue, with in the text `This is item "Fix xxxxx" from the list in #399`, and track your progress there. We'll update the list to link this new issue in the list.

I don't feel like it's necessary to open issues for an item in the list until someone decides to work on it. What do you think?

**Comment by p-l-** [comment]
@gpotter2 thanks a lot for this! It's fine to let it as an issue open by you (I can edit the text anyway) since that's your work. You can create a PR to reference this URL (`https://github.com/secdev/scapy/issues/399`) in `CONTRIBUTING.md`, that's an excellent idea. Thanks again!

**Comment by xee5ch** [comment]
Sounds good to me.  I was only suggesting setting up different issues to gain attention of people with projects like [First Timers Only](http://drakvuf.com/) and a  small list of other intiatives in this place.

I want to help you the way you want, I will start with low-hanging fruit.  I only started using Scapy for the first time in course 4-6 months ago, wish to get back into Python, and you cannot take an infosec course without your project coming up.  I just want to give back.

**Comment by p-l-** [comment]
@xee5ch OK so maybe we could open (and reference) issues only for issues we consider important, some easy, some medium, and some more difficult? I really like the idea of estimating the difficulty.

**Comment by xee5ch** [comment]
Sure.  I am not so familiar with the project.  I can spot the ones I understand (trivial), and the others I don't understand.  How can I help you with this?  Since I am not a project member, I cannot label, right?

**Comment by p-l-** [comment]
If @gpotter2 agrees with my proposal you can create a new issue for something and mention this issue. We'll update the list and label it "help wanted". But I'd like first to know what @gpotter2 thinks about it, since that's his idea and work.

**Comment by gpotter2** [comment]
I totally agree with you :) 
https://github.com/secdev/scapy/pull/402

**Comment by gpotter2** [comment]
Ups

**Comment by guedou** [comment]
Wow, that's a lot of work for 2017 =)

Could you add somes lines related to PR and issues still waiting on bitbucket ?

**Comment by gpotter2** [comment]
@guedou You mean all of them ? Which one are relevent ?

**Comment by p-l-** [comment]
@gpotter2 I think @guedou means that if some of the items also have opened issues and/or pull request on Bitbucket that you are aware of, maybe we could add a link to that as resources.

**Comment by guedou** [comment]
Indeed, I only meant to specify that people might want to have a look at open PR and issues on bitbucket, as a reminder. A single line is more than enough.

They are some PRs that are worth looking at.

**Comment by p-l-** [comment]
@guedou text updated. I'll remove some comments here to leave the issue as clean as possible, if you don't mind.

**Comment by gpotter2** [comment]
Added a line at the bottom... @p-l- you may remove useless/outdated comments

**Comment by xee5ch** [comment]
I am not sure if I am beating a dead horse, but do we need someone with lesser experience, perhaps myself, to attempt a cleanup of Bitbucket issues and move what's need over here, close BB issues as pointed and extant on GH, or should we leave it be for now?

**Comment by p-l-** [comment]
@xee5ch Thank you but we have decided to leave the Bitbucket issues and PR for now, and only close those tracked or fixed here. I think your time is better used trying to fix easy issues for now, so that you understand Scapy's internals. That's also probably more interesting & fun.

If you do find an issue or a PR on Bitbucket that you think is still useful and valid, you're welcome to open an issue or a PR here to track it, reference the Bitbucket issue / PR, then close it on Bitbucket.

**Comment by plorinquer** [comment]
@gpotter2 Small typo (I guess): "Finish and add length of packet in [vtp.py](https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/contrib/vtp.py#L85)" instead of [ntp.py](https://github.com/secdev/scapy/blob/cce88a93b006775957757bd2d311649f4d423863/scapy/contrib/vtp.py#L85)

**Comment by gpotter2** [comment]
@plorinquer thx

**Comment by plorinquer** [comment]
@gpotter2 I hope this update is enough to consider that these have been fixed : #461

> Fix cdp.py here and here

**Comment by gpotter2** [comment]
@plorinquer Definitely :smile:

**Comment by brianbsquared** [comment]
I'd like to pick up "Handle properly the error in this in fields.py". Is raising a Scapy_Exception the correct approach for this?

**Comment by Chibbluffy** [comment]
Please add IPFix layer as work in progress. It will be in a new file ipfix.py
Thanks

**Comment by adharshkamath** [comment]
@gpotter2 and anyone concerned, 
I'm currently working on developing a module to create NDN packets similar to the IP packets module in Scapy. Would you be interested in merging that with Scapy?

**Comment by guedou** [comment]
All contributions are welcomed if they conform to the contributing.md file.

**Comment by p-l-** [comment]
Also @adharshkamath, you can usually reach people on the [gitter](https://gitter.im/secdev/scapy).

**Comment by Asmeble** [comment]
this needs to be fixed immediately.
